### ISSUE TITLE
map_entry causes unable to infer None type

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9973]
### Summary

.

### Lint Name

map_entry

### Reproducer

I tried this code:
apparently this was an official test at some point :upside_down_face: 
```rust
#![allow(unused, clippy::needless_pass_by_value)]
use std::collections::{BTreeMap, HashMap};
use std::hash::Hash;

fn insert_if_absent2<K: Eq + Hash, V>(m: &mut HashMap<K, V>, k: K, v: V) {
    if !m.contains_key(&k) { m.insert(k, v) } else { None };
}

fn main() {}
```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::map_entry`
```
warning: failed to automatically apply fixes suggested by rustc to crate `f`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
 --> src/main.rs:7:5
  |
7 |     None } else { None };
  |     ^^^^ cannot infer type of the type parameter `T` declared on the enum `Option`
  |
help: consider specifying the generic argument
  |
7 |     None::<T> } else { None };
  |         +++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: usage of `contains_key` followed by `insert` on a `HashMap`
 --> src/main.rs:6:5
  |
6 |     if !m.contains_key(&k) { m.insert(k, v) } else { None };
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
  = note: requested on the command line with `-W clippy::map-entry`
help: try this
  |
6 ~     if let std::collections::hash_map::Entry::Vacant(e) = m.entry(k) { e.insert(v);
7 ~     None } else { None };
  |

warning: `f` (bin "f") generated 1 warning (run `cargo fix --bin "f"` to apply 1 suggestion)
warning: `f` (bin "f" test) generated 1 warning (1 duplicate)
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (1eb62b123 2022-11-27)
binary: rustc
commit-hash: 1eb62b1235fd77200e6bd967d70e83c0f2497233
commit-date: 2022-11-27
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
